### PR TITLE
ramips: add support for Telewell TW-4G (LTE) router

### PR DESCRIPTION
**Pull request by JiiPee74**
- 2.4GHz wifi working ok
- 5GHz wifi shows up but power is very weak
- WPA3 security will put this hardware on it's knees, better to use WPA2
- This device is clone of Lava lr25g001 router

### CODE REVIEW

**Comment by mrkiko** [comment]
But I don't know if this is still relevant in current mt76 code.

File: target/linux/ramips/mt7620/base-files/etc/board.d/01_leds [hunk 1]
```diff
@@ -143,9 +143,10 @@ kingston,mlwg2|\
 sanlinking,d240)
 	ucidef_set_led_netdev "wifi_led" "wifi" "blue:wifi" "wlan0"
 	;;
-lava,lr-25g001)
-	ucidef_set_led_netdev "wlan2g" "WiFi 2.4GHz" "green:wlan2g" "wlan1"
-	ucidef_set_led_netdev "wlan5g" "WiFi 5GHz" "green:wlan5g" "wlan0"
+lava,lr-25g001|\
+telewell,tw-4g-lte)
+	ucidef_set_led_netdev "wifi2g" "WiFi 2.4GHz" "green:wifi2g" "wlan1"
+	ucidef_set_led_netdev "wifi5g" "WiFi 5GHz" "green:wifi5g" "wlan0"
```
[the remark this answers is not in the document]
Comment (JiiPee74) [reply target/linux/ramips/mt7620/base-files/etc/board.d/01_leds:149]: They are clone devices, thats why I didn't separate them. But if you want, I can separate them. Or maybe @CHKDSK88 could test this if Lava router is working with these changes?

[the remark this answers is not in the document]
Comment (s-2) [reply target/linux/ramips/mt7620/base-files/etc/board.d/01_leds:149]: > this device and Lava router are clones, my .dts is quite much copy paste of Lava .dts

So this should be split into a common `.dtsi` and two `.dts` for the devices?
Separation and changes to the other device should still be in a separate commit though.

[the remark this answers is not in the document]
Comment (adschm) [reply target/linux/ramips/mt7620/base-files/etc/board.d/01_leds:149]: Generally:
If you change settings for other devices, please do it separately.

Concretely:
This is indeed broken for the lava device. Nevertheless, wifi LEDs should be implemented with DT triggers anyway, so you would just not touch this file at all for your new device.

File: target/linux/ramips/mt7620/base-files/etc/board.d/02_network [hunk 2]
```diff
@@ -173,9 +173,10 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"1:lan:4" "2:lan:3" "3:lan:2" "4:lan:1" "0:wan" "6@eth0"
 		;;
-	lava,lr-25g001)
+	lava,lr-25g001|\
+	telewell,tw-4g-lte)
 		ucidef_add_switch "switch0" \
-			"1:lan" "2:lan" "3:lan" "4:lan" "5:wan" "0@eth0"
+			"1:lan:4" "2:lan:3" "3:lan:2" "4:lan:1" "5:wan" "0@eth0"
```
Comment (adschm) [inline target/linux/ramips/mt7620/base-files/etc/board.d/02_network:179]: Separate device, separate change.

Comment (JiiPee74) [reply target/linux/ramips/mt7620/base-files/etc/board.d/02_network]: Devices are now separated. Only jboot generic stuff is piled as one. Hope this is finaly ok for everybody.
Should I make pr to fix Lava router leds config or does someone else do it?

File: target/linux/ramips/mt7620/base-files/etc/board.d/01_leds (hunk 1 above)
[the remark this answers is not in the document]
Comment (JiiPee74) [reply target/linux/ramips/mt7620/base-files/etc/board.d/01_leds]: Devices are now separated

File: target/linux/ramips/mt7620/base-files/etc/board.d/02_network (hunk 2 above)
Comment (adschm) [reply target/linux/ramips/mt7620/base-files/etc/board.d/02_network]: If you resolve the question, it's hard for people to answer it.

But yes, the split was what I asked for regarding network config.

**Comment by adschm** [comment]
Please use DT triggers for this device. If you have it, it should be easy to test.
Example: https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7620a_tplink_archer-c20-v1.dts#L48

Those have the advantage that the LED and trigger are linked directly.
The current implementation will actually break immediately when somebody renames the WiFi interfaces, so it shouldn't be used anymore.

You will either use phy0tpt and phy1tpt or phy0radio and phy1radio. The former includes activity, but is not available on all devices.
https://openwrt.org/docs/guide-user/base-system/led_configuration#wifi_activity_triggers

File: target/linux/ramips/dts/mt7620a_telewell_tw-4g-lte.dts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
[the remark this answers is not in the document]
Comment (JiiPee74) [reply target/linux/ramips/dts/mt7620a_telewell_tw-4g-lte.dts]: @adschm I have now run this router some time with this new flash clock setting, seems to work. If there ain't anything else requiring changes, this PR is done from my part and I will put this router in use, so after few days it won't be available for testing anymore.

Thanks.

**Comment by JiiPee74** [comment]
ffs! I am so sick of git because I don't know it well enough. Now this PR is totally broken because I managed to do something on git desktop client, forgot that branch I was messing was bind to this PR... :(
How I can make this shit undone?

**Comment by mans0n** [comment]
I'm not sure if it's relevant to you but this might help you:
https://openwrt.org/docs/guide-developer/working-with-github-pr#reopen_closed_pr

**Comment by JiiPee74** [comment]
> I'm not sure if it's relevant to you but this might help you: https://openwrt.org/docs/guide-developer/working-with-github-pr#reopen_closed_pr The key point is that you can checkout to the commit and relocate the branch.

Luckily I had in my build VM branch unchanged so I got it fixed with force push.

**Comment by mans0n** [comment]
And please squash your commits.
https://openwrt.org/docs/guide-developer/working-with-github-pr#squashing_commits

**Comment by CHKDSK88** [comment]
> > Does all Ethernet ports working?
> 
> Yes, with #4303 what is already in master, all switch ports are working fine. You Still got a Lava router? Could you test doing build if everything works fine on Lava also? I did rename leds because they wasn't working + fixed router ports naming because they was backward. I tought that it's most likely same issue on Lava router as they are clones after all.

I checked ports and it was fixed in: https://github.com/openwrt/openwrt/commit/a82fa5bcd4817339f5eb75cbeb49e1ab6d30d50e

Are You sure, if after restart OpenWrt, every port have autoneg enabled and doesn't need fix from a82fa5bcd4817339f5eb75cbeb49e1ab6d30d50e?


Please rebase Your PR to current master: 02_network was changed, ane Your entry is duplicate:
https://github.com/openwrt/openwrt/blob/a82fa5bcd4817339f5eb75cbeb49e1ab6d30d50e/target/linux/ramips/mt7620/base-files/etc/board.d/02_network#L140

**Comment by JiiPee74** [comment]
> > > Does all Ethernet ports working?
> > 
> > 
> > Yes, with #4303 what is already in master, all switch ports are working fine. You Still got a Lava router? Could you test doing build if everything works fine on Lava also? I did rename leds because they wasn't working + fixed router ports naming because they was backward. I tought that it's most likely same issue on Lava router as they are clones after all.
> 
> I checked ports and it was fixed in: [a82fa5b](https://github.com/openwrt/openwrt/commit/a82fa5bcd4817339f5eb75cbeb49e1ab6d30d50e)
> 
> Are You sure, if after restart OpenWrt, every port have autoneg enabled and doesn't need fix from [a82fa5b](https://github.com/openwrt/openwrt/commit/a82fa5bcd4817339f5eb75cbeb49e1ab6d30d50e)?
> 
> Please rebase Your PR to current master: 02_network was changed, ane Your entry is duplicate:
> 
> https://github.com/openwrt/openwrt/blob/a82fa5bcd4817339f5eb75cbeb49e1ab6d30d50e/target/linux/ramips/mt7620/base-files/etc/board.d/02_network#L140

Read previous comments. I was specially ASKED to make it duplicate, even that I didn't want to do it because it is clone device. But I do what ´masters ask me to do, so it is not separated and duplicate for that reason.

**Comment by CHKDSK88** [comment]
@JiiPee74 
> 
> Read previous comments. I was specially ASKED to make it duplicate, even that I didn't want to do it because it is clone device. But I do what ´masters ask me to do, so it is now separated and duplicate for that reason.

I'm pretty sure, that @adschm  had problem with differences. Your device has exactly  the same network layout like `engenius,esr600 `. In `02_network` file, duplicates are not welcome.

**Comment by JiiPee74** [comment]
> I'm pretty sure, that @adschm had problem with differences. Your device has exactly the same network layout like `engenius,esr600 `. In `02_network` file, duplicates are not welcome.

I was trying to push that change for lava network port naming, I did even ping you if you could have test if it works and because I do not own lava device, big bosses had issue with that change and I had to make it separate and now it's duplicate.

But I dont have much motivation to make adjustment this anymore, because this will never gets supported unless someone finds workaround for JBOOT 2MB kernel limit.
I have understood that next major openwrt release will be based to 5.10.x kernel and I cannot build 5.10.x kernel for this device  because 2MB image limit.
And this would get into 21.02 only if this device was in high demand, what it's not.

So unless there is workaround, anything I do for this PR is quite much just waste of my time.

**Comment by CHKDSK88** [comment]
Here is solution of 2MB kernel problem: https://github.com/openwrt/openwrt/pull/4255
It wait for merge.

**Comment by JiiPee74** [comment]
> Here is solution of 2MB kernel problem: #4255 It wait for merge.

Ah nice, maybe there is some hope then after all. I see this is quite recent work from you guys, no wonder I haven't spotted it. Maybe I'll test it at some point but right now I have this router in use so I don't really wanna do much if any testing with it.

**Comment by mpratt14** [comment]
@JiiPee74 now that ramips is on kernel 5.10 and the other fixes are done, you may want to try again here

also, put flash chip model in commit message (you can find it in kernel log)

**Pull request closed without merging**